Prioritize task execution by recipe order (#4)
ESMValGroup/ESMValCore: doc/esmvalcore/recipe.rst

The diagnostics section includes one or more diagnostics. Each diagnostics will
include:

- the variables to preprocess, including the preprocessor to be applied to each variable;
- the diagnostic script(s) to be run;
- a description of the diagnostic and lists of themes and realms that it applies to;
- an optional ``additional_datasets`` section.

The diagnostic section defines tasks
------------------------------------

> The diagnostic section(s) define the tasks that will be run when running the recipe.

Reviewer: ```suggestion
The diagnostic section defines the tasks that will be executed when running the recipe.
```

For each variable a preprocesing task will be defined and for each diagnostic script a
diagnostic task will be defined. If variables need to be derived
from other variables, a preprocessing task for each of the variables
needed to derive that variable will be defined as well. These tasks can be viewed

> in the main_log_debug.txt file that is produced every run. Each task has a unique

Reviewer: ```suggestion
in the main_log_debug.txt file that is produced by every run. Each task has a unique
```

> name that defines the directory where the results of that task are stored. Task

Reviewer: ```suggestion
name that defines the subdirectory where the results of that task are stored in `work_dir`. Task
```

A (simplified) example diagnostics section could look like

.. code-block:: yaml

  diagnostics:
    diagnostic_name:
      description: Air temperature tutorial diagnostic.
      themes:
        - phys
      realms:
        - atmos
      variables:
        variable_name:
          short_name: ta
          preprocessor: preprocessor_name
          mip: Amon
      scripts:
        script_name:
          script: examples/diagnostic.py

Note that the example recipe above contains a single diagnostic section
called ``diagnostic_name`` and will result in two tasks:

- a preprocessing task called ``diagnostic_name/variable_name`` that will preprocess
  air temperature data for each dataset in the Datasets_ section of the recipe (not shown).
- a diagnostic task called ``diagnostic_name/script_name``

The path to the script provided in the `script` option should be
either the absolute path to the script, or the path relative to the
``esmvaltool/diag_scripts`` directory.

Ancestor tasks
--------------
Some tasks require the result of other tasks to be ready before they can start,
e.g. a diagnostic script needs the preprocessed variable data to start. Thus

> each tasks has zero or more ancestor tasks. By the default, each diagnostic task

Reviewer: ```suggestion
each tasks has zero or more ancestor tasks. By default, each diagnostic task
```

          short_name: pr
          preprocessor: preprocessor_name
          mip: Amon
      scripts:
        script_b:
          script: diagnostic_b.py
          ancestors: [diagnostic_1/script_a, precip]

The example recipe above will result in four tasks:

- a preprocessing task called ``diagnostic_1/airtemp``
- a diagnostic task called ``diagnostic_1/script_a``
- a preprocessing task called ``diagnostic_2/precip``
- a diagnostic task called ``diagnostic_2/script_b``

the preprocessing tasks do not have any ancestors, while the diagnostic_a.py
script will receive the preprocessed air temperature data
(has ancestor ``diagnostic_1/airtemp``) and the diagnostic_b.py
script will receive the results of diagnostic_a.py and the preprocessed precipitation
data (has ancestors ``diagnostic_1/script_a`` and ``diagnostic_2/precip``).

Task priority
-------------
Tasks are assigned a priority, with tasks appearing earlier on in the recipe
getting higher priority. The tasks will be executed sequentially or in parellel,
depending on the setting of `max_parallel_tasks` in the :ref:`user configuration file`.
When choosing the next task to execute, the task with the highest priority that

> is not waiting for ancestor tasks will be chosen. This feature makes it possible to

Reviewer: ```suggestion
is not waiting for ancestor tasks will be started. This feature makes it possible to
```
